hsm/m0hsm.c

```diff
@@ -0,0 +1,748 @@
+	m0_obj_init(&obj, &uber_realm, &id, m0_client_layout_id(instance));
+
+	/* open the entity */
+	rc = open_entity(&obj.ob_entity);
+	if (rc) {
+		fprintf(stderr,
+			"open object %#"PRIx64":%#"PRIx64" failed: %d\n",
+			id.u_hi, id.u_lo, rc);
+		return rc;
+	}
+
+	while ((read_nr = read(fd, io_buff, IO_SIZE)) > 0) {
+		rc = m0hsm_pwrite(&obj, io_buff, read_nr, off);
+		if (rc) {
+			fprintf(stderr, "error on writting at %zu to "
+				"%#"PRIx64":%#"PRIx64": %d\n", off,
+				id.u_hi, id.u_lo, rc);
+			break;
+		}
+
+		/* increase offset */
+		off += read_nr;
+	}
+
+	if (read_nr < 0) {
+		fprintf(stderr, "error after reading %zu bytes from '%s': %s\n",
+			off, path, strerror(errno));
+		rc = -1;
+	}
+
+	m0_entity_fini(&obj.ob_entity);
+	return rc;
+}
+
+
+static int load_env(char **tgt, const char *env)
+{
+	*tgt = getenv(env);
+	if (*tgt == NULL) {
+		fprintf(stderr, "%s environment variable is not set\n", env);
+		return -1;
+	}
+	return 0;
+}
+
+static int load_client_env()
+{
+	if (load_env(&local_addr, "CLIENT_LADDR"))
+		return -1;
+	if (load_env(&ha_addr, "CLIENT_HA_ADDR"))
+		return -1;
+	if (load_env(&profile, "CLIENT_PROF_OPT"))
+		return -1;
+	if (load_env(&proc_fid, "CLIENT_PROC_FID"))
+		return -1;
+	return 0;
+}
+
+static void usage()
+{
+	printf("Usage: m0hsm <action> <fid> [...]\n");
+	printf("  actions:\n");
+	printf("    create <fid> <tier>\n");
+	printf("    show <fid>\n");
+	printf("    dump <fid>\n");
+	printf("    write <fid> <offset> <len> <seed>\n");
+	printf("    write_file <fid> <path>\n");
+	printf("    read <fid> <offset> <len>\n");
+	printf("    copy <fid> <offset> <len> <src_tier> "
+			"<tgt_tier> [options: mv,keep_prev,w2dest]\n");
+	printf("    move <fid> <offset> <len> <src_tier> "
+			"<tgt_tier> [options: keep_prev,w2dest]\n");
+	printf("    stage <fid> <offset> <len> <tgt_tier> "
+			"[options: mv,w2dest]\n");
+	printf("    archive <fid> <offset> <len> <tgt_tier> "
+			"[options: mv,keep_prev,w2dest]\n");
+	printf("    release <fid> <offset> <len> <tier> "
+			"[options: keep_latest]\n");
+	printf("    multi_release <fid> <offset> <len> <max_tier> "
+			"[options: keep_latest]\n");
+	printf("    set_write_tier <fid> <tier>\n\n");
+	printf("  <fid> parameter format is [hi:]lo. "
+	                  "(hi == 0 if not specified.)\n");
+	printf("  The numbers are read in decimal, hexadecimal "
+	                              "(when prefixed with `0x')\n"
+	       "  or octal (when prefixed with `0') formats.\n");
+}
+
+
+/* read a 64 bits argument. Interprets 0x as base16, 0 as base 8,
+ * others as base 10 */
+int64_t read_arg64(const char *s)
+{
+	long long ll = -1LL;
+
+	if (sscanf(s, "%lli", &ll) != 1)
+		return -1LL;
+
+	return ll;
+}
+
+/**
+ * Read FID in the format [hi:]lo.
+ * Return how many numbers are read.
+ */
+int read_fid(const char *s, struct m0_uint128 *fid)
+{
+	int res;
+	long long hi, lo;
+
+	res = sscanf(s, "%lli:%lli", &hi, &lo);
+	if (res == 1) {
+		fid->u_lo = hi;
```

Comment:
need to set:
```
fid->u_hi = 0;
```
?

Reply:
Yes, let's set it to be on the safe side.